init kineto after torch module initialized

staugust opened the pull request:
Fixes #131020 

As discussed in the issue thread,  we can use ` KINETO_DAEMON_INIT_DELAY_S` to delay the initialization of `kineto`  in case `kineto` is initialized before `libtorch_cuda.so`. 

It's not clear to set a proper value of environmental variable `KINETO_DAEMON_INIT_DELAY_S`, here's a trick to make the initialization of `kineto` after the initialization of module `torch`. I'm not sure whether this is an acceptable trick, please take a look at this pr, thanks.

staugust committed:
init kineto after torch module initialized

sraikund16 commented:
This seems reasonable to me. @briancoutinho do you see any issue with initializing the client later in the stub? 

Also @staugust do you mind adding a test to make sure KINETO_DAEMON_INIT_DELAY_S is no longer necessary?

sraikund16 commented on torch/csrc/profiler/kineto_client_interface.cpp:
```diff
@@ -91,26 +100,6 @@ struct RegisterLibKinetoClient {
-    if (std::getenv("KINETO_USE_DAEMON") != nullptr) {
-      int init_delay_s = get_init_delay();
-      if (init_delay_s > 0) {
-        std::thread t([init_delay_s, kineto_init]() {
```
If we are removing init_delay_s we should remove KINETO_DAEMON_INIT_DELAY_S from all areas of the code

staugust replied:
Thanks, this initial commit is a proposal to show another solution for kineto initialization.   As it's reasonable, I will continue improve this pr as comments advised.

sraikund16 reviewed (changes requested):
Please check comments above

staugust commented:
> This seems reasonable to me. @briancoutinho do you see any issue with initializing the client later in the stub?
> 
> Also @staugust do you mind adding a test to make sure KINETO_DAEMON_INIT_DELAY_S is no longer necessary?

OK，I'd like to add a testcase to check pytorch on-demand profiling to ensure it.

staugust committed:
add testcase for kineto on-demand profiling, check #131020

staugust committed:
format kineto test code with lintrunner

staugust commented:
/easycla

staugust commented:
@aaronenyeshi  @briancoutinho @sraikund16  would you like to spend a few minutes to review this pr?  Thanks.

staugust commented:
> > This seems reasonable to me. @briancoutinho do you see any issue with initializing the client later in the stub?
> > Also @staugust do you mind adding a test to make sure KINETO_DAEMON_INIT_DELAY_S is no longer necessary?
> 
> OK，I'd like to add a testcase to check pytorch on-demand profiling to ensure it.

Testcase is added #132375

staugust commented:
@pytorchbot label "topic: not user facing

staugust commented:
@pytorchbot label "topic: not user facing"

staugust commented:
@aaronenyeshi @briancoutinho @sraikund16 , hello, it's been a while since this PR was last updated. Could you please find some time to review this PR when you're available? Thank you in advance.

briancoutinho commented:
@staugust changes look good to me :)

briancoutinho commented:
@pytorchmergebot merge

briancoutinho commented:
@pytorchmergebot merge

sraikund16 commented:
@pytorchbot merge

sraikund16 commented:
@pytorchbot rebase

sraikund16 commented on torch/csrc/stub.c:
```diff
@@ -9,7 +9,11 @@ extern "C"
 __attribute__((visibility("default"))) PyObject* PyInit__C(void);
 #endif
 
+extern "C" TORCH_API void global_kineto_init();
```
I don't think you need the extern "C" here that is for cpp only

staugust replied:
Yeah, that's true,  'extern' is enough for function declaration in c files.

sraikund16 reviewed (changes requested):
I think we are using extern "C" wrong here. Please use the suggestion in this post: https://stackoverflow.com/a/2403522

staugust committed:
declare global_kineto_init in correct way, and do not init kineto as ondemand tracing is not supported on Apple or edge platform

sraikund16 commented on torch/csrc/stub.c:
```diff
@@ -9,7 +9,11 @@ extern "C"
 __attribute__((visibility("default"))) PyObject* PyInit__C(void);
 #endif
 
+extern void global_kineto_init();
```
I think we might need to move this prototype to a header for when stub.c is not compiled with kineto_client_interface

staugust replied:
Yes, this prototype can be moved to a header file. By moving this prototype to a header, do you mean some macros should be added to determine whether `stub.c` is compiled with `kineto_client_interface`?  Here's a simple example: 

```
#ifndef TORCH_KINETO_H
#define TORCH_KINETO_H

#ifndef USE_KINETO 
// stub.c is compiled with `kineto_client_interface`, `torch._C` can find symbol `global_kineto_init`. 
void global_kineto_init();
#else
// stub.c is not compiled with `kineto_client_interface`,  `PyMODINIT_FUNC PyInit__C(void)` calls a empty function. 
inline void global_kineto_init(){}
#endif
#endif // TORCH_KINETO_H
```

sraikund16 replied:
I might need to look at the compile paths but I just see this PR failing a lot of builds if you look below

sraikund16 replied:
Based on the failures it is failing because of missing prototype in kineto_client_interface so we need to have a header that this file and stub.c includes to make the builds compile

staugust replied:
Sorry, I was out of office last week. I'd check those builds and try to fix it today.

staugust replied:
I created a new header file `kineto_client_interface.h`, and declare `global_kineto_init` inside this header file. Then,  I moved the call to `global_kineto_init` into `torch/csrc/Module.cpp:initModule` instead of `stub.c` as `torch/csrc/Module.cpp` has already included header files under `torch/csrc/profiler`, that is `#include <torch/csrc/profiler/combined_traceback.h>`.

cyyever commented on torch/csrc/profiler/kineto_client_interface.cpp:
```diff
@@ -71,46 +71,24 @@ class LibKinetoClient : public libkineto::ClientInterface {
-    return std::stoi(delay_s);
-  } catch (const std::invalid_argument& _) {
-    return -1;
+  if (std::getenv("KINETO_USE_DAEMON") != nullptr) {
```
Use thread-safe c10::utils::get_env

staugust replied:
OK, just updated the code, could you please approve the workflows to check whether this update is ok?

staugust committed:
remove TORCH_API from function prototype

From the definition of `TORCH_API` in `c10/macros/Export.h`, we can see
that `TORCH_API` is designed to decorate functions in `libtorch.so`.

staugust committed:
init kineto inside initModule

staugust committed:
fixup head file name in kineto_client_interface.cpp

staugust committed:
fixup: use c10::utils::get_env instead of std::get_env

staugust committed:
add namespace for global_kineto_init

staugust commented:
I've synced my forked branch with the latest main branch, the two failed `linux-focal-cuda11.8-py3.10-gcc9/test` distributed cases should be solved.

staugust committed:
format kineto_client_interface.h

cyyever commented:
@staugust 
```
ImportError: /var/lib/jenkins/.cache/bazel/_bazel_jenkins/fdf6d09bf4b4f04a71e2a7dfceb40620/sandbox/processwrapper-sandbox/6490/execroot/pytorch/bazel-out/k8-fastbuild/bin/test_bazel.runfiles/pytorch/torch/_C.so: undefined symbol: _ZN5torch18global_kineto_initEv
```
Should fix it.

cyyever commented on torch/csrc/Module.cpp:
```diff
@@ -2443,6 +2444,8 @@ Call this whenever a new thread is created in order to propagate values from
   torch::set_disabled_torch_dispatch_impl(
       PyObject_GetAttrString(module, "_disabled_torch_dispatch_impl"));
   ASSERT_TRUE(torch::disabled_torch_dispatch_impl() != nullptr);
```
#ifdef USE_KINETO

cyyever replied:
I think it should be wrapped inside #ifdef USE_KINETO

staugust committed:
wrap global_kineto_init with macro USE_KINETO

staugust commented:
> @staugust
> 
> ```
> ImportError: /var/lib/jenkins/.cache/bazel/_bazel_jenkins/fdf6d09bf4b4f04a71e2a7dfceb40620/sandbox/processwrapper-sandbox/6490/execroot/pytorch/bazel-out/k8-fastbuild/bin/test_bazel.runfiles/pytorch/torch/_C.so: undefined symbol: _ZN5torch18global_kineto_initEv
> ```
> 
> Should fix it.

Yes, maybe the macro `USE_KIENTO` can solve this problem. I'm trying to set up an environment to re-produce this error, but I cannot download the docker image `308535385114.dkr.ecr.us-east-1.amazonaws.com/pytorch/pytorch-linux-focal-cuda12.1-cudnn9-py3-gcc9:5b7480b1a7b6b3ac2be441917d8079acf9f5994c` used in job `linux-focal-cuda12.1-py3.10-gcc9-bazel-test `.  @cyyever  @sraikund16  Do you know how can I pull this docker image ?  Thanks.

cyyever commented:
@staugust 
Build the code using 
```
env USE_KINETO=0 python setup.py
```

cyyever commented:
@pytorchbot rebase -b  main

cyyever commented:
@pytorchbot merge

staugust commented:
> @staugust Build Torch by
> 
> ```
> env USE_KINETO=0 python setup.py
> ```

I think those checks will cover this case, `USE_KINETO=0`. I've built pytorch with `#ifdef USE_KINETO`, it works as well.  Now, I'm working on build a docker image to re-produce  `linux-focal-cuda12.4-py3.10-gcc9-bazel-test`.
